jmh-benchmarks/src/main/java/org/apache/kafka/jmh/util/ConcurrentMapBenchmark.java

```diff
@@ -0,0 +1,175 @@
+    private Map<Integer, Integer> copyOnWriteMap;
+    private int writePerLoops;
+
+    @Setup
+    public void setup() {
+        Map<Integer, Integer> mapTemplate = IntStream.range(0, mapSize).boxed()
+                .collect(Collectors.toMap(i -> i, i -> i));
+        concurrentHashMap = new ConcurrentHashMap<>(mapTemplate);
+        copyOnWriteMap = new CopyOnWriteMap<>(mapTemplate);
+        writePerLoops = TIMES / (int) Math.round(writePercentage * TIMES);
+    }
+
+    @Benchmark
+    @OperationsPerInvocation(TIMES)
+    public void testConcurrentHashMapGet(Blackhole blackhole) {
+        for (int i = 0; i < TIMES; i++) {
+            if (i % writePerLoops == 0) {
+                // add offset mapSize to ensure computeIfAbsent do add new entry
+                concurrentHashMap.computeIfAbsent(i + mapSize, key -> key);
+            } else {
+                blackhole.consume(concurrentHashMap.get(i % mapSize));
+            }
+        }
+    }
+
+    @Benchmark
+    @OperationsPerInvocation(TIMES)
+    public void testConcurrentHashMapGetRandom(Blackhole blackhole) {
+        for (int i = 0; i < TIMES; i++) {
+            if (i % writePerLoops == 0) {
+                // add offset mapSize to ensure computeIfAbsent do add new entry
+                concurrentHashMap.computeIfAbsent(i + mapSize, key -> key);
+            } else {
+                blackhole.consume(concurrentHashMap.get(ThreadLocalRandom.current().nextInt(0, mapSize + 1)));
+            }
+        }
+    }
+
+    @Benchmark
+    @OperationsPerInvocation(TIMES)
+    public void testCopyOnWriteMapGet(Blackhole blackhole) {
+        for (int i = 0; i < TIMES; i++) {
+            if (i % writePerLoops == 0) {
+                // add offset mapSize to ensure computeIfAbsent do add new entry
+                copyOnWriteMap.computeIfAbsent(i + mapSize, key -> key);
+            } else {
+                blackhole.consume(copyOnWriteMap.get(i % mapSize));
+            }
+        }
+    }
+
+    @Benchmark
+    @OperationsPerInvocation(TIMES)
+    public void testCopyOnWriteMapGetRandom(Blackhole blackhole) {
+        for (int i = 0; i < TIMES; i++) {
+            if (i % writePerLoops == 0) {
+                // add offset mapSize to ensure computeIfAbsent do add new entry
+                copyOnWriteMap.computeIfAbsent(i + mapSize, key -> key);
+            } else {
+                blackhole.consume(copyOnWriteMap.get(ThreadLocalRandom.current().nextInt(0, mapSize + 1)));
+            }
+        }
+    }
+
+    @Benchmark
+    @OperationsPerInvocation(TIMES)
+    public void testConcurrentHashMapValues(Blackhole blackhole) {
+        for (int i = 0; i < TIMES; i++) {
+            if (i % writePerLoops == 0) {
+                // add offset mapSize to ensure computeIfAbsent do add new entry
+                concurrentHashMap.computeIfAbsent(i + mapSize, key -> key);
+            } else {
+                blackhole.consume(concurrentHashMap.values());
```

Comment:
maybe we need to iterate it. the time complexity can be very different

Reply:
Thanks, the benchmark result are as belows

```
Benchmark                                              (mapSize)  (writePercentage)  Mode  Cnt      Score     Error  Units
ConcurrentMapBenchmark.testConcurrentHashMapEntrySet         100                0.1  avgt   15  30675.669 ± 305.105  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGet              100                0.1  avgt   15      2.494 ±   0.024  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGetRandom        100                0.1  avgt   15      5.134 ±   0.083  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapValues           100                0.1  avgt   15  33012.029 ± 677.201  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapEntrySet            100                0.1  avgt   15  34543.961 ± 398.451  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGet                 100                0.1  avgt   15      2.845 ±   0.041  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGetRandom           100                0.1  avgt   15      4.218 ±   0.062  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapValues              100                0.1  avgt   15  26139.956 ± 229.991  ns/op
```

Reply:
It is weird ... why `CopyOnWrite#entrySet` is slower but  `CopyOnWrite#Values` is not ?

Reply:
It is a bug, setup level should be Invocation instead of the default value (Trial), otherwise computeIfAbsent won't work in the iterations other than the first iteration.

Reply:
Now the benchmark result are as belows

```text
Benchmark                                              (mapSize)  (writePercentage)  Mode  Cnt      Score      Error  Units
ConcurrentMapBenchmark.testConcurrentHashMapEntrySet         100                0.1  avgt    5  22338.964 ± 3148.123  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGet              100                0.1  avgt    5     13.175 ±    0.488  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGetRandom        100                0.1  avgt    5     16.382 ±    1.301  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapValues           100                0.1  avgt    5  19684.342 ± 2892.526  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapEntrySet            100                0.1  avgt    5  21260.254 ±  860.984  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGet                 100                0.1  avgt    5   5436.128 ±  883.432  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapGetRandom           100                0.1  avgt    5   5606.463 ±  962.939  ns/op
ConcurrentMapBenchmark.testCopyOnWriteMapValues              100                0.1  avgt    5  16483.967 ± 1072.444  ns/op
```

Reply:
It seems the cost of `testCopyOnWriteMapGet` is the "write" need to do deep-copy, right? 

Maybe we can revise the use case of `RecordAccumulator#nodeStats`. It updates rarely as the number of nodes won't be changed frequently. Also, it call `get` only.

Reply:
Make sense to me, should I open a JIRA or just create a minor pr for this one ?

Reply:
> Make sense to me, should I open a JIRA or just create a minor pr for this one ?

yep. BTW, could you please add comments to `CopyOnWriteMap` to describe this benchmark?

Reply:
filed https://issues.apache.org/jira/browse/KAFKA-16942